### ISSUE TITLE
Failed to instantiate smart contract written by GO

### ISSUE THREAD

**Issue by SipengXie** [IBM-Blockchain/blockchain-vscode-extension#3007]
OS : Win 10
Version of the extension : v2.0.2
Path of my project : $GOPATH
My project provides go.mod go.sum and I ran go mod vendor before packaging it.

The error is like:
`Failed to install on peer Org1 Peer with reason: Could not install smart contact received error: failed with status:500 ::failed to invoke backing implementation of 'InstallChaincode': could not build chaincode: external builder failed: external builder failed to build: external builder 'golang' failed: exit status 1`

**Comment by ryu-os** [comment]
The 'Tutorial Gallery' only has instructions for a 'Typescript Smart Contract'. Is there a step-by-step tutorial for installing/instantiating a 'Go Smart Contract' with the prerequisites listed?

**Comment by james-wallis** [comment]
Hi, I've had some time to test this today. I did the following:

1. Created new local 1 org environment
2. Create Golang smart contract (using V2)
3. Packaged it (using V2)
4. Deployed it onto my local environment
5. Created an asset and read it back

I'm on an Intel Mac and on the latest (V2.0.2) version.

I wonder if this is a Windows issue but I don't have a machine available to test?
Also are you able to test following the method I've used?

**Comment by Childult** [comment]
> Hi, I've had some time to test this today. I did the following:
> 
> 1. Created new local 1 org environment
> 2. Create Golang smart contract (using V2)
> 3. Packaged it (using V2)
> 4. Deployed it onto my local environment
> 5. Created an asset and read it back
> 
> I'm on an Intel Mac and on the latest (V2.0.2) version.
> 
> I wonder if this is a Windows issue but I don't have a machine available to test?
> Also are you able to test following the method I've used?

I ran the method you mentioned on WSL2 (Ubuntu-20.04) and the error occurs:
`Failed to install on peer Org1 Peer with reason: Could not install smart contact received error: failed with status:500 ::failed to invoke backing implementation of 'InstallChaincode': could not build chaincode: external builder failed: external builder failed to build: external builder 'golang' failed: exit status 1`
I ran `go mod vendor` but it doesn't help.
I also tried the javascript smart contract. It works well.

**Comment by paulatwork** [comment]
The IBM Blockchain Platform extension is pretty cool. And I am very happy to get the default GOLANG Smart Contract working! 

1. Use GO to fix the module dependencies.
2. Use the Hyperledger Fabric binaries (peer) to package the GO SmartContract. 

**MY FULLY WORKED EXAMPLE .. MY INFO:**
OS : macOS Big Sur 11.6
Version of the extension : v2.0.4
GO Version: 1.17.2
GOPATH="/Users/paul/dev/go"
GOBIN="/Users/paul/dev/go/bin"
GOROOT="/usr/local/go"

My smart contact name = "smartcontract05"
My path to IBM Blockchain Platform Smart Contract project: $GOPATH/src/blockchain/smartcontract05

**Prerequisite**: Ensure Go is installed and working correctly. 
a. Ensure GO installed correctly.
b. Ensure GO 'modules' working correctly with path etc. For example, refer 'Call your code from another module' - https://golang.org/doc/tutorial/call-module-code.

**Prerequisite** Optional: Setup VS Code GO Extension.
a. Nice to have .. but not essential, install 'Go for Visual Studio Code' extension.
b. Disable 'gopls' to fix some bugs so the GO SmartContract code has including IntelliSense and code navigation features. Set "go.useLanguageServer": false. VSCode > Code > Preferences > Settings - Search 'go.useLanguageServer'. Uncheck.

**Using IBM Blockchain Platform Extension - Create GO chaincode with smart contract**

A2.3: Smart Contracts - "Create New Project" 
A2.4: Accept 'Default Contract type'
A2.5: Smart Contract Language: Select 'Go'

For example, my version is installed into directory '$GOPATH/src/blockchain/smartcontract05'

NOTE: The GO template code uses V2 Hyperledger 'GO contract API'. 

**Install Hyperledger Fabric (a) binaries, (b) Fabric Samples files, & (c) Docker Images**

Info: https://hyperledger-fabric.readthedocs.io/en/release-2.2/install.html
(Note, if Docker is not active during the download script, the Hyperledger Docker Images not downloaded)

1. Create a new location on your machine where you want to place the fabric-samples repository and enter that directory in a terminal window. Run: `mkdir /Users/paul/dev/hyperledgerfabric`
2. Change to the new directory. Run: `cd /Users/paul/dev/hyperledgerfabric`
3. Clone hyperledger/fabric-samples repo and latest release of the binaries & images. Run: `curl -sSL https://bit.ly/2ysbOFE | bash -s`
4. Set the FABRIC_CFG_PATH to point to the core.yaml file in the fabric-samples repository. Run: `export FABRIC_CFG_PATH=/Users/paul/dev/hyperledgerfabric/fabric-samples/config`
5. (optional) Check peer version. Run: `/Users/paul/dev/hyperledgerfabric/fabric-samples/bin/peer version`

**Using Hyperledger 'peer' binaries - Package the GO Smart Contact**
Info: https://hyperledger-fabric.readthedocs.io/en/release-2.2/deploy_chaincode.html

1. Change back to the smartconract directory. Run: `cd $GOPATH/src/blockchain/smartcontract05`
2. Prepackage. Update the default 'mod' file to clean up unused dependencies or add missing dependencies in the SmartContract mod file. Run: `go mod tidy`
3. Prepackage. Download package dependencies (constructs a directory named vendor in the main module's root directory that contains copies of all packages needed to support builds and tests of packages in the main module). Run: `GO111MODULE=on go mod vendor`
4. Package Smart Contract chaincode (into tar.gz'). For my example, run: `/Users/paul/dev/hyperledgerfabric/fabric-samples/bin/peer lifecycle chaincode package smartcontract05.tar.gz --path . --lang golang --label smartcontract05_1.0`

**Using IBM Blockchain Platform Extension - Deploy Smart Contract**

A3.5: Smart Contracts view - **Import** a Package
A3.6: Select "tar.gz" file created by the 'peer' binaries.
A3.7: .. continue as per Tutorial directions.

Work like a charm ! Yay.

**Comment by winterpi** [comment]
Any updates?
I also run into this problem.
And I've already followed @[paulatwork](https://github.com/paulatwork) 's suggestions, but it still can not install chaincode. 
BTW, when Package Smart Contract chaincode, it pointed out
```
Fatal error when initializing core config : error when reading core config file: Config File "core" Not Found in "[/home/winter/go/src/github.com/winter/my-chaincode]"
```
unless I config FABRIC_CFG_PATH in ~/.bashrc file for all terminals.

**Comment by winterpi** [comment]
aha, I got the solution. 

Check the go version in the go.mod file of chaincode. 
Make sure that it should **be consistent with** that of your plugin environment. 

E.g.:  
in fabric-samples/chaincode/fabcar/go sample, the version required in go.mod is ```go 1.13``` 
change it to ```go 1.18```, which is the go version of my desktop. 
Then ```GO111MODULE=on go mod vendor``` again , and package it, install will be OK
